src/librbd/api/Migration.cc

```diff
@@ -337,6 +339,28 @@ int open_source_image(librados::IoCtx& io_ctx, const std::string &image_name,
   return 0;
 }
 
+class SteppedProgressContext : public ProgressContext {
+public:
+  SteppedProgressContext(ProgressContext* progress_ctx, size_t total_steps)
+    : m_progress_ctx(progress_ctx), m_total_steps(total_steps) {
+  }
+
+  void next_step() {
+    ++m_current_step;
+  }
+
+  int update_progress(uint64_t object_number,
+                      uint64_t object_count) override {
+    return m_progress_ctx->update_progress(
+      object_number * m_current_step, object_count * m_total_steps);
```

Comment:
shouldn't it be `(m_current_step - 1) * object_count + object_number` instead of `object_number * m_current_step`?